src/hypothesis/searchstrategy/domains.py

```diff
@@ -0,0 +1,116 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2016 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import re
+import inspect
+from pathlib import Path
+from encodings import idna
+
+import hypothesis.strategies as st
+from hypothesis.internal.compat import to_unicode
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+MAX_LABEL_SIZE = 63
+MAX_LABEL_COUNT = 127
+MAX_DOMAIN_SIZE = 253
+
+# Each label must be 1 < 63 octets/bytes long,
+# the number of labels must be 0 < 127 and
+# the total length must be < 253
+
+class DomainStrategy(SearchStrategy):
+
+    """A strategy for domain name strings, defined in terms of lists of text
+    strings."""
+
+
+    def __init__(self, alphabet, example_domains_only):
+        SearchStrategy.__init__(self)
+        self.label_text = st.text(
+            min_size=1, max_size=MAX_LABEL_SIZE, alphabet=self.alphabet
+        )
+
+        if example_domains_only:
+            self.top_level = st.sampled_from(EXAMPLE_DOMAINS)
+        else:
+            self.top_level = st.sampled_from(SUFFIX_LIST)
+
+    @staticmethod
+    def canonicalize(string):
+        return idna.ToASCII(idna.nameprep(to_unicode(string)))
+
+    def do_draw(self, data):
+        top_level = data.draw(self.top_level)
+        total = len(top_level) + 1 # One extra for the dot
+
+        labels = [] # RFC term for subdomains
+        label_count = data.draw(
+            st.integers(min_value=1, max_value=MAX_LABEL_COUNT)
+        ) - top_level.count(".")
+
+
+        while total < MAX_DOMAIN_SIZE and len(labels) < label_count:
+            try:
+                label = self.canonicalize(data.draw(self.label_text))
+            except UnicodeError:
+                continue
```

Comment:
There should be some sort of fail safe here to avoid infinite loops, but I'm not sure how to go about doing it. Just use a counter? Should it be reset if it succeeds or not?